docs/community/20241112-node-bootstrapping.md

```diff
@@ -0,0 +1,111 @@
+cleanly integrate my contributions into the node bootstrapping process without
+interfering with other implementations.
+I would like my contributions to be as re-usable as possible across different
+bootstrap providers.
+I would like to read and to follow documentation, guidelines, and specifications
+on the above.
+I would like to offer a choice of node bootstrapping configuration systems to users,
+enable distribution vendors to extend the choces available by adding new
+configuration systems, and ease maintenance and enable extension of currently
+supported systems for the ClusterAPI developer community.
+
+
+**Problem statement / Example issues**
+
+Currently, the kubeadm bootstrap provider implements two provisioning systems:
+cloud-init and ignition.
+As there is currently no abstract way to pass configuration to  node bootstrapping,
+both configuration system implementations - ignition and cloud-init - must cover
+generic configuration as well as config language specific concerns.
+This leads to duplicate code and to maintenance overhead.
+Also, other providers like e.g. MicroK8s must implement their own OS configuration
+generators since kubeadm's node bootstrapping implementations cannot be re-used.
+MicroK8s currently only supports cloud-init.
+
+Outside of ClusterAPI, both [cloud-init](https://github.com/canonical/cloud-init)
+and [ignition](https://github.com/coreos/ignition) provisioning are widely adopted
+across distributions.
+While cloud-init is used by general-purpose Linux distributions like Ubuntu/Debian,
+SUSE Linux, Alma, Rocky, Fedora, and Red Hat Enterprise Linux, ignition is popular
+with special-purpose distributions like Fedora CoreOS / Red Hat CoreOS, SuSE MicroOS
+/ SLE Micro / Kalpa, and Flatcar Container Linux.
+It is likely that more provisioning systems exist; a clean separation between bootstrap and provisioning
+and guidelines on how each is developed will make it easier to add support to ClusterAPI as well as to
+share implementations across bootstrap providers.
+
+
+**Compatibility and API guarantees**
+
+ClusterAPI has been stable for multiple years and is in widespread production use.
+Proposals and implementations of this working group must ensure to not excert operational
+risk on existing integrations.
+Any changes in the kubeadm bootstrap provider in particular must uphold guarantees and
+expectations set by the current implementation and must continue to support current use
+cases OR must provide means for current use cases to continue functioning for an appropriate
+migration period, as well as define a clear path for migration.
+
+## Scope
+
+1. Propose architectural improvements that abstract system configuration with the goal of
+   reducing duplicate functionality in current implementations, easing development of new configuration
+   features, and simplifying integration of new configuration systems for node bootstrapping.
```

Comment:
I'm also interested in this idea in general (primarily adding a comment to this spot to have a thread). We're currently using metal3 to do bare metal provisioning, and have started work on a new bare metal oriented provider, both of which have a similar problem that most others do not: deploying outside of a cloud platform that take over most of the basic configuration.

To apply custom configuration it's still most convenient to use configuration systems like cloud-init etc., as that makes it a lot easier for users to provide custom images. This leads to two problems that are related to the scope of this WG:
- While systems like cloud-init allow you to provide "vendor" configuration in addition to the user configuration, the options for the former are somewhat limited (don't know about Ignition). In our experience it was necessary to use the bootstrap provider in order to apply configuration files, e.g. for advanced networking setups. This breaks the separation between infrastructure and Kubernetes configuration, and gets even more complex and tangled up when mixing in variables provided on a per-machine basis (possible with metal3 for example).
- Bare metal providers (or providers for platforms that do not come with enough automation) need to implement explicit support for configuration systems in order to use them to apply their configuration. If the system doesn't support providing multiple configurations (e.g. the inconsistent vendor/user config in cloud-init), the infra provider needs to merge configuration, making it even more complicated.

Both of these could be improved by allowing to extend the abstract configuration before turning it into a system-specific format.

I'm not sure if this topic came up already (I haven't read all the comments here or on the google doc), but I think this could be a good opportunity to make some improvements in CAPI geared towards on-prem use.

cc @kashifest - in case you're interested from a metal3 perspective

Reply:
This sounds great - and I second the sentiment of making this a focus item in the working group. I've added a "Maintainability, Backwards Compatibility, and Extensibility" track to the WG agenda: https://docs.google.com/document/d/12v6NFr7xal9RH3GEB_1IR6HENq2GMuzALS3l5elvrMg/edit?usp=sharing and took the liberty of adding you as a point of contact for these concerns.

Will close this thread to ease review of this PR as aligned on in the [Nov 27 office hours](https://docs.google.com/document/d/1GgFbaYs-H6J5HSQ6a7n4aKpk0nDLE2hgG2NSOM9YIRw/edit?tab=t.0#heading=h.faoavlb0ym66).

Reply:
(Oh and @schrej  I've added you to the WG proposal as a Stakeholder, please let me know if that works for you)

Reply:
I totally agree with @schrej's point of view here